File: lemur-py/kots.py
"""
KOTS: Key-Homomorphic One-Time Signature.

All randomness is derived from SHAKE XOFs:
  - Public matrix A has the structured form A = [I; A2] and only A2 is
    expanded from SHAKE128(seed || [i,j] || b'A').
  - Secret keys are sampled per-entry from SHAKE256(seed || [i,j] || b'S').
  - Message hash H(mu) uses SHAKE256(mu || index || b'H').
"""

import numpy as np
from Crypto.Hash import SHAKE128, SHAKE256

from profiles import LemurProfile, DEFAULT
from ring import Ring
from sample import (
    GaussianSampler,
    build_cdt,
    xof_gauss_poly,
    xof_ternary_poly,
    xof_uniform_poly,
)


class KOTS:
    """Key-Homomorphic One-Time Signature scheme.

    Default profile: profiles.DEFAULT (d256_k4 — d=256, tau=20, N=1024).
    Individual keyword overrides take precedence over profile fields,
    which is how `make_codec(tau=...)` keeps working.
    """

    def __init__(self, profile: LemurProfile | None = None, *,
                 d=None, q=None, k=None, ell=None, m=None, n=None,
                 alpha=None, alpha_h=None, beta_z=None, beta_sigma=None,
                 lam=128, sigma: float | None = None,
        sampler: GaussianSampler | None = None):
        if profile is None:
            profile = DEFAULT
        self.profile    = profile
        self.d          = d          if d          is not None else profile.d
        self.q          = q          if q          is not None else profile.q_prime
        self.k          = k          if k          is not None else profile.k
        self.ell        = ell        if ell        is not None else profile.ell
        self.m          = m          if m          is not None else profile.m
        self.n          = n          if n          is not None else profile.n
        self.alpha      = alpha      if alpha      is not None else profile.alpha
        self.alpha_h    = alpha_h    if alpha_h    is not None else profile.alpha_h
        self.beta_z     = beta_z     if beta_z     is not None else profile.beta_z
        self.beta_sigma = beta_sigma if beta_sigma is not None else profile.beta_sigma
        self.lam        = lam
        # `alpha` is the paper symbol (Gaussian width parameter; appears in
        # norm-bound derivations).  The sampler's sigma is the actual CDT
        # stddev: sigma = alpha / sqrt(2*pi).  Both come from the profile
        # unless the caller overrides.
        if sampler is None:
            sigma_val = profile.sigma if sigma is None else sigma
            sampler = GaussianSampler(
                sigma=sigma_val,
                cdt_bits=profile.cdt_bits,
                tailcut=profile.tailcut,
            )
        self.sampler    = sampler
        # Derived bit widths (used by codec)
        self.logq       = (self.q - 1).bit_length()
        self.bits_z     = (2 * self.beta_z).bit_length()
        self.bits_zagg  = (2 * self.beta_sigma).bit_length()
        # Ring with NTT
        self.ring       = Ring(self.q, self.d)
        # CDT table for discrete Gaussian keygen (built once at init)
        self.cdt        = build_cdt(
            self.sampler.sigma,
            self.sampler.cdt_bits,
            self.sampler.tailcut,
        )

    @staticmethod
    def _inf_norm(Z: np.ndarray) -> int:
        """Max absolute coefficient of a polynomial or polynomial matrix."""
        return int(np.max(np.abs(Z)))

    # -----------------------------------------------------------------------
    # Hash: mu -> H(mu) in T_{alpha_h}
    # -----------------------------------------------------------------------

    def _hash_to_ternary(self, mu, index=0):
        xof = SHAKE256.new(mu + index.to_bytes(4, 'little') + b'H')
        return xof_ternary_poly(xof, self.alpha_h, self.d)

    def _build_H(self, mu):
        """Construct H = [I_ell | H'] in R^{ell x k}.  Returns shape (ell, k, d)."""
        k, ell, d = self.k, self.ell, self.d
        H = np.zeros((ell, k, d), dtype=np.int64)
        for i in range(ell):
            H[i, i, 0] = 1
        for i in range(ell):
            for j in range(k - ell):
                H[i, ell + j] = self._hash_to_ternary(mu, index=i * (k - ell) + j)
        return H

    # -----------------------------------------------------------------------
    # KOTS algorithms
    # -----------------------------------------------------------------------

    def setup(self, seed):
        """Setup(seed) -> A2 in R_q^{(m-n) x n}.

        The effective public matrix is the structured matrix
            A = [I_n; A2] in R_q^{m x n},
        but only the lower block A2 is materialized and stored.
        """
        rows = self.m - self.n
        A2 = np.zeros((rows, self.n, self.d), dtype=np.int64)
        for i in range(rows):
            for j in range(self.n):
                xof = SHAKE128.new(seed + bytes([i, j]) + b'A')
                A2[i, j] = xof_uniform_poly(xof, self.q, self.d)
        return A2

    def _matmul_with_A(self, X, A2):
        """Multiply X by the implicit structured matrix A = [I; A2] mod q."""
        top = X[:, :self.n].copy() % self.q
        bottom = self.ring.mat_mul(X[:, self.n:], A2)
        return (top + bottom) % self.q

    def keygen(self, A2, seed):
        """KGen(A2, seed) -> (S, T).

        S[i][j] from SHAKE256(seed || [i,j] || b'S'), T = S * A mod q where
        A is the implicit structured matrix [I; A2].
        """
        S = np.zeros((self.k, self.m, self.d), dtype=np.int64)
        for i in range(self.k):
            for j in range(self.m):
                xof = SHAKE256.new(seed + bytes([i, j]) + b'S')
                S[i, j] = xof_gauss_poly(xof, self.cdt, self.sampler.cdt_bits, self.d)
        T = self._matmul_with_A(S, A2)
        return S, T

    def sign(self, A2, S, mu):
        """Sign(A2, S, mu) -> Z = H * S in R^{ell x m}.

        Each Z[i,j] = sum_l H[i,l] * S[l,j] (over Z, exact signed arithmetic).
        The coefficient bound k * alpha_h * max|S| << q/2 so mul_signed is valid.
        """
        H = self._build_H(mu)
        Z = np.zeros((self.ell, self.m, self.d), dtype=np.int64)
        for i in range(self.ell):
            for j in range(self.m):
                for l in range(self.k):
                    Z[i, j] = Z[i, j] + self.ring.mul_signed(H[i, l], S[l, j])
        return Z

    def vrfy(self, A2, T, mu, Z, beta):
        """Vrfy: return True iff ||Z||_inf <= beta and Z*A == H*T (mod q)."""
        if self._inf_norm(Z) > beta:
            return False
        H  = self._build_H(mu)
        ZA = self._matmul_with_A(Z, A2)
        HT = self.ring.mat_mul(H, T)
        return bool(np.all(ZA == HT))

    def ivrfy(self, A2, T, mu, Z):
        return self.vrfy(A2, T, mu, Z, self.beta_z)

    def svrfy(self, A2, T, mu, Z):
        return self.vrfy(A2, T, mu, Z, self.beta_sigma)

    def wvrfy(self, A2, T, mu, Z):
        return self.vrfy(A2, T, mu, Z, 2 * self.beta_sigma)


# ---------------------------------------------------------------------------
# Quick correctness test
# Module-level alias so hvc / lemur / cli can keep `from kots import inf_norm`.
inf_norm = KOTS._inf_norm


# ---------------------------------------------------------------------------

if __name__ == "__main__":
    kots = KOTS()
    print(f"KOTS: d={kots.d}, q={kots.q}, k={kots.k}, ell={kots.ell}, "
          f"m={kots.m}, n={kots.n}")
    print(f"  alpha={kots.alpha}, alpha_h={kots.alpha_h}, "
          f"beta_z={kots.beta_z}, beta_sigma={kots.beta_sigma}")
    print(f"  logq={kots.logq}, bits_z={kots.bits_z}, bits_zagg={kots.bits_zagg}")
    print(f"  ring zeta={kots.ring.zeta}")
    print()

    A2   = kots.setup(bytes(range(32)))
    S, T = kots.keygen(A2, bytes(range(32, 64)))
    mu   = b"test message"
    Z    = kots.sign(A2, S, mu)

    print(f"  implicit A shape: ({kots.m}, {kots.n}), stored A2 shape: {A2.shape[:2]}")
    print(f"  ||Z||_inf = {kots._inf_norm(Z)}  (beta_z={kots.beta_z})")
    print(f"  ivrfy: {'PASS' if kots.ivrfy(A2, T, mu, Z) else 'FAIL'}")
    print(f"  svrfy: {'PASS' if kots.svrfy(A2, T, mu, Z) else 'FAIL'}")
    Z2 = kots.sign(A2, S, b"wrong")
    print(f"  wrong-msg ivrfy: "
          f"{'FAIL (expected)' if not kots.ivrfy(A2, T, mu, Z2) else 'PASS (unexpected!)'}")

File: lemur-rs/src/kots.rs
//! KOTS: Key-Homomorphic One-Time Signature.
//!
//! The shipped parameter set satisfies the proof condition `q' ≡ 17 (mod 32)`,
//! which forbids a length-d negacyclic NTT in `R_{q'}`.  Multiplication
//! is therefore routed through the CRT-via-aux-primes backend in
//! [`crate::aux_ntt`]: two 48-bit auxiliary NTT-friendly primes whose
//! product is large enough to recover the exact signed product before
//! reducing mod q'.
//!
//! The signing path is constant-time.

use sha3::digest::XofReader;
use zeroize::{Zeroize, ZeroizeOnDrop};

use crate::aux_ntt::CrtBackend;
use crate::error::LemurError;
use crate::profile::Profile;
use crate::sample::{
    kots_hash_xof, kots_keygen_xof, kots_setup_xof, xof_gauss_poly_ctx_into,
    xof_gauss_poly_with_profile_fill, xof_ternary_poly, xof_uniform_poly, GaussCtx,
};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// KOTS public matrix A2.
///
/// Shape: `(m-n) x n x d` coefficient-domain, flat `Vec<i64>` in `[0, q')`.
/// The pre-NTT'd companion stores, per polynomial, its forward NTT under
/// each of the two auxiliary primes — so a matmul against `A2` costs only
/// one forward NTT pair per input operand plus one inverse NTT pair and a
/// CRT combine per output polynomial (vs three per output for naive `mul`).
///
/// `ntt_p1`, `ntt_p2` are flat `(m-n) * n * d` `u64` buffers; entry `(i, j)`
/// starts at offset `(i * n + j) * d` and runs for `d` words.
#[derive(Clone)]
pub struct KotsA {
    pub coeffs: Vec<i64>,
    pub ntt_p1: Vec<u64>,
    pub ntt_p2: Vec<u64>,
}

/// KOTS secret key S: shape (k x m x d), flat i64 array.
#[derive(Clone, Zeroize, ZeroizeOnDrop)]
pub struct KotsSk(pub Vec<i64>);

/// KOTS public key T: shape (k x n x d), flat i64 array.
#[derive(Clone)]
pub struct KotsPk(pub Vec<i64>);

/// KOTS signature Z: shape (ell x m x d), flat i64 array.
#[derive(Clone)]
pub struct KotsSig(pub Vec<i64>);

/// Maximum absolute coefficient of a polynomial array.
pub fn inf_norm(v: &[i64]) -> i64 {
    v.iter().map(|&x| x.abs()).max().unwrap_or(0)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/// Build the pre-NTT caches for an `A2` matrix already in coefficient form.
fn build_a2_ntt_cache(
    coeffs: &[i64],
    a2_rows: usize,
    n: usize,
    backend: &CrtBackend,
) -> (Vec<u64>, Vec<u64>) {
    let d = backend.d();
    let mut p1 = vec![0u64; a2_rows * n * d];
    let mut p2 = vec![0u64; a2_rows * n * d];
    for i in 0..a2_rows {
        for j in 0..n {
            let off = (i * n + j) * d;
            backend.forward_into_i64(
                &coeffs[off..off + d],
                &mut p1[off..off + d],
                &mut p2[off..off + d],
            );
        }
    }
    (p1, p2)
}

/// Multiply X (rows x m) by the structured A = [I_n; A2].
///
/// Caches X's NTT pair once per (row, left-col) pair and accumulates
/// pointwise products in each auxiliary prime's NTT domain.
fn matmul_structured_a(x: &[i64], rows: usize, a: &KotsA, profile: &Profile) -> Vec<i64> {
    let m = profile.m;
    let n = profile.n;
    let a2_rows = m - n;
    let cfg = profile.kots_crt().expect("CRT KOTS profile");
    // Each output coefficient sums `a2_rows` per-coefficient products
    // (X[:, n:][i,k] · A2[k,j] for k in 0..a2_rows) in the paired NTT
    // domain before reconstruction; size the backend for that.
    let backend = cfg.backend_for_accum(a2_rows);
    let d = cfg.d;
    let q = cfg.q as i64;

    // Forward-NTT X[:, n:] into the paired aux-prime domain once.  Layout
    // mirrors `a.ntt_p{1,2}`: (rows x a2_rows) polynomials at stride d.
    let mut xr_p1 = vec![0u64; rows * a2_rows * d];
    let mut xr_p2 = vec![0u64; rows * a2_rows * d];
    for i in 0..rows {
        for k in 0..a2_rows {
            let src_off = (i * m + n + k) * d;
            let dst_off = (i * a2_rows + k) * d;
            backend.forward_into_i64(
                &x[src_off..src_off + d],
                &mut xr_p1[dst_off..dst_off + d],
                &mut xr_p2[dst_off..dst_off + d],
            );
        }
    }

    // Accumulate T_right[i, j] = sum_k X[:, n:][i, k] * A2[k, j] in NTT domain.
    let mut t = vec![0i64; rows * n * d];
    let mut acc_p1 = vec![0u64; d];
    let mut acc_p2 = vec![0u64; d];
    for i in 0..rows {
        for j in 0..n {
            acc_p1.iter_mut().for_each(|v| *v = 0);
            acc_p2.iter_mut().for_each(|v| *v = 0);
            for k in 0..a2_rows {
                let xr_off = (i * a2_rows + k) * d;
                let a2_off = (k * n + j) * d;
                backend.accum_mul_slices(
                    &mut acc_p1,
                    &mut acc_p2,
                    &xr_p1[xr_off..xr_off + d],
                    &xr_p2[xr_off..xr_off + d],
                    &a.ntt_p1[a2_off..a2_off + d],
                    &a.ntt_p2[a2_off..a2_off + d],
                );
            }
            let prod = backend.finalize_accum_slices(&mut acc_p1, &mut acc_p2);
            // T[i, j] = X[:, :n][i, j] + T_right[i, j] (mod q').
            for l in 0..d {
                let x_val = x[(i * m + j) * d + l];
                let p_val = prod[l] as i64;
                t[(i * n + j) * d + l] = (x_val + p_val).rem_euclid(q);
            }
        }
    }
    t
}

/// Multiply the structured KOTS hash matrix `H = [I_ell | H']` by `B`.
///
/// This is the only CRT matrix multiply needed on the sign / verify paths.
/// The identity block contributes `B[i, j]` directly in coefficient form,
/// so we only forward-NTT and multiply the hashed columns `ell..k`.
fn mat_mul_h(
    h: &[i64],
    b: &[i64],
    ell: usize,
    k: usize,
    t: usize,
    backend: &CrtBackend,
) -> Vec<i64> {
    let d = backend.d();
    let q = backend.q() as i64;
    let hashed_cols = k - ell;

    debug_assert_eq!(h.len(), ell * k * d);
    debug_assert_eq!(b.len(), k * t * d);

    // Pre-NTT only H' and the matching lower rows of B.  The identity
    // columns H[i,i] = 1 are copied directly into the coefficient-domain
    // output below.
    let mut h_p1 = vec![0u64; ell * hashed_cols * d];
    let mut h_p2 = vec![0u64; ell * hashed_cols * d];
    for i in 0..ell {
        for l in 0..hashed_cols {
            let h_col = ell + l;
            let src_off = (i * k + h_col) * d;
            let dst_off = (i * hashed_cols + l) * d;
            backend.forward_into_i64(
                &h[src_off..src_off + d],
                &mut h_p1[dst_off..dst_off + d],
                &mut h_p2[dst_off..dst_off + d],
            );
        }
    }

    let mut b_p1 = vec![0u64; hashed_cols * t * d];
    let mut b_p2 = vec![0u64; hashed_cols * t * d];
    for l in 0..hashed_cols {
        let b_row = ell + l;
        for j in 0..t {
            let src_off = (b_row * t + j) * d;
            let dst_off = (l * t + j) * d;
            backend.forward_into_i64(
                &b[src_off..src_off + d],
                &mut b_p1[dst_off..dst_off + d],
                &mut b_p2[dst_off..dst_off + d],
            );
        }
    }

    let mut out = vec![0i64; ell * t * d];
    let mut acc_p1 = vec![0u64; d];
    let mut acc_p2 = vec![0u64; d];
    for i in 0..ell {
        for j in 0..t {
            acc_p1.iter_mut().for_each(|v| *v = 0);
            acc_p2.iter_mut().for_each(|v| *v = 0);
            for l in 0..hashed_cols {
                let h_off = (i * hashed_cols + l) * d;
                let b_off = (l * t + j) * d;
                backend.accum_mul_slices(
                    &mut acc_p1,
                    &mut acc_p2,
                    &h_p1[h_off..h_off + d],
                    &h_p2[h_off..h_off + d],
                    &b_p1[b_off..b_off + d],
                    &b_p2[b_off..b_off + d],
                );
            }

            let out_off = (i * t + j) * d;
            let identity_off = (i * t + j) * d;
            if hashed_cols == 0 {
                for l in 0..d {
                    out[out_off + l] = b[identity_off + l].rem_euclid(q);
                }
                continue;
            }

            let prod = backend.finalize_accum_slices(&mut acc_p1, &mut acc_p2);
            for l in 0..d {
                out[out_off + l] = (b[identity_off + l] + prod[l] as i64).rem_euclid(q);
            }
        }
    }
    out
}

/// Build H = [I_ell | H'] in R^{ell x k}, stride = profile.d.
fn build_h(mu: &[u8], profile: &Profile) -> Vec<i64> {
    let ell = profile.ell;
    let k = profile.k;
    let alpha_h = profile.alpha_h;
    let d = profile.d;

    let mut h = vec![0i64; ell * k * d];
    for i in 0..ell {
        h[(i * k + i) * d] = 1;
    }
    for i in 0..ell {
        for j in 0..(k - ell) {
            let index = (i * (k - ell) + j) as u32;
            let mut xof = kots_hash_xof(mu, index);
            let poly = xof_ternary_poly(&mut xof as &mut dyn XofReader, alpha_h, d);
            let col = ell + j;
            h[(i * k + col) * d..(i * k + col + 1) * d].copy_from_slice(&poly);
        }
    }
    h
}

// ---------------------------------------------------------------------------
// Setup / KeyGen / Sign / Verify
// ---------------------------------------------------------------------------

/// KOTS Setup: expand A2 from seed and pre-NTT it under both aux primes.
pub fn kots_setup(seed: &[u8], profile: &'static Profile) -> KotsA {
    let cfg = profile.kots_crt().expect("CRT KOTS profile");
    let backend = cfg.backend();
    let d = cfg.d;
    let q = cfg.q;
    let m = profile.m;
    let n = profile.n;
    let a2_rows = m - n;

    let mut coeffs = vec![0i64; a2_rows * n * d];
    for i in 0..a2_rows {
        for j in 0..n {
            let mut xof = kots_setup_xof(seed, i, j);
            let poly = xof_uniform_poly(&mut xof as &mut dyn XofReader, q, d);
            coeffs[(i * n + j) * d..(i * n + j + 1) * d].copy_from_slice(&poly);
        }
    }
    let (ntt_p1, ntt_p2) = build_a2_ntt_cache(&coeffs, a2_rows, n, &backend);
    KotsA {
        coeffs,
        ntt_p1,
        ntt_p2,
    }
}

/// KOTS KeyGen using the profile's default Gaussian sampler.
pub fn kots_keygen(a: &KotsA, seed: &[u8], profile: &'static Profile) -> (KotsSk, KotsPk) {
    let d = profile.d;
    let k = profile.k;
    let m = profile.m;

    let mut s = vec![0i64; k * m * d];
    for i in 0..k {
        for j in 0..m {
            let mut xof = kots_keygen_xof(seed, i, j);
            xof_gauss_poly_with_profile_fill(
                &mut xof as &mut dyn XofReader,
                profile,
                &mut s[(i * m + j) * d..(i * m + j + 1) * d],
            );
        }
    }
    let t = matmul_structured_a(&s, k, a, profile);
    (KotsSk(s), KotsPk(t))
}

/// KOTS KeyGen with a caller-supplied Gaussian sampler context (used by the
/// bench harness to swap CDT widths / sigmas without rebuilding the baked
/// table).
pub fn kots_keygen_ctx(
    a: &KotsA,
    seed: &[u8],
    profile: &'static Profile,
    ctx: &GaussCtx,
) -> (KotsSk, KotsPk) {
    let d = profile.d;
    let k = profile.k;
    let m = profile.m;

    let mut s = vec![0i64; k * m * d];
    for i in 0..k {
        for j in 0..m {
            let mut xof = kots_keygen_xof(seed, i, j);
            xof_gauss_poly_ctx_into(
                &mut xof as &mut dyn XofReader,
                ctx,
                &mut s[(i * m + j) * d..(i * m + j + 1) * d],
            );
        }
    }
    let t = matmul_structured_a(&s, k, a, profile);
    (KotsSk(s), KotsPk(t))
}

/// KOTS pk-only KeyGen — avoids returning the full secret key when only
/// the public key is needed (used by HVC leaf expansion).
pub fn kots_pk_from_seed(a: &KotsA, seed: &[u8], profile: &'static Profile) -> KotsPk {
    let d = profile.d;
    let k = profile.k;
    let m = profile.m;

    let mut s = vec![0i64; k * m * d];
    for i in 0..k {
        for j in 0..m {
            let mut xof = kots_keygen_xof(seed, i, j);
            xof_gauss_poly_with_profile_fill(
                &mut xof as &mut dyn XofReader,
                profile,
                &mut s[(i * m + j) * d..(i * m + j + 1) * d],
            );
        }
    }
    let t = matmul_structured_a(&s, k, a, profile);
    KotsPk(t)
}

/// KOTS Sign: Z = H * S with signed lift.
pub fn kots_sign(sk: &KotsSk, mu: &[u8], profile: &'static Profile) -> KotsSig {
    let cfg = profile.kots_crt().expect("CRT KOTS profile");
    let ell = profile.ell;
    let k = profile.k;
    let m = profile.m;
    // H = [I_ell | H']; only the hashed `k - ell` columns require ring
    // products.  Size the backend for that lazy accumulation depth.
    let backend = cfg.backend_for_accum(k - ell);
    let q = cfg.q as i64;
    let h = build_h(mu, profile);

    let z_unsigned = mat_mul_h(&h, &sk.0, ell, k, m, &backend);
    let half = q / 2;
    let z: Vec<i64> = z_unsigned
        .into_iter()
        .map(|x| if x > half { x - q } else { x })
        .collect();
    KotsSig(z)
}

/// KOTS Verify: ||Z||_inf <= beta AND Z*A == H*T (mod q').
pub fn kots_vrfy(
    a: &KotsA,
    pk: &KotsPk,
    mu: &[u8],
    sig: &KotsSig,
    beta: i64,
    profile: &'static Profile,
) -> Result<(), LemurError> {
    if inf_norm(&sig.0) > beta {
        return Err(LemurError::VerifyFailed);
    }
    let cfg = profile.kots_crt().expect("CRT KOTS profile");
    let ell = profile.ell;
    let k = profile.k;
    let n = profile.n;
    let backend = cfg.backend_for_accum(k - ell);

    let h = build_h(mu, profile);
    let za = matmul_structured_a(&sig.0, ell, a, profile);
    let ht = mat_mul_h(&h, &pk.0, ell, k, n, &backend);
    if za == ht {
        Ok(())
    } else {
        Err(LemurError::VerifyFailed)
    }
}

/// Individual-verify bound β_z.
pub fn kots_ivrfy(
    a: &KotsA,
    pk: &KotsPk,
    mu: &[u8],
    sig: &KotsSig,
    profile: &'static Profile,
) -> Result<(), LemurError> {
    kots_vrfy(a, pk, mu, sig, profile.beta_z, profile)
}

/// Strong-verify bound β_σ.
pub fn kots_svrfy(
    a: &KotsA,
    pk: &KotsPk,
    mu: &[u8],
    sig: &KotsSig,
    profile: &'static Profile,
) -> Result<(), LemurError> {
    kots_vrfy(a, pk, mu, sig, profile.beta_sigma, profile)
}

/// Weak-verify bound 2·β_σ.
pub fn kots_wvrfy(
    a: &KotsA,
    pk: &KotsPk,
    mu: &[u8],
    sig: &KotsSig,
    profile: &'static Profile,
) -> Result<(), LemurError> {
    kots_vrfy(a, pk, mu, sig, 2 * profile.beta_sigma, profile)
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::profile::D256_K4;

    fn run_self_consistency(profile: &'static Profile) {
        let seed_setup = [0x42u8; 32];
        let seed_key = [0x43u8; 32];
        let mu: &[u8] = b"kots self-consistency";

        let a = kots_setup(&seed_setup, profile);
        assert_eq!(
            a.coeffs.len(),
            (profile.m - profile.n) * profile.n * profile.d
        );
        assert_eq!(a.ntt_p1.len(), a.coeffs.len());
        assert_eq!(a.ntt_p2.len(), a.coeffs.len());
        let q_kots = profile.q_kots() as i64;
        assert!(a.coeffs.iter().all(|&x| (0..q_kots).contains(&x)));

        let (sk, pk) = kots_keygen(&a, &seed_key, profile);
        let pk_only = kots_pk_from_seed(&a, &seed_key, profile);
        assert_eq!(pk.0, pk_only.0, "pk_from_seed drift vs keygen");

        assert_eq!(sk.0.len(), profile.k * profile.m * profile.d);
        assert_eq!(pk.0.len(), profile.k * profile.n * profile.d);

        let sig = kots_sign(&sk, mu, profile);
        assert_eq!(sig.0.len(), profile.ell * profile.m * profile.d);
        kots_ivrfy(&a, &pk, mu, &sig, profile).expect("ivrfy");
        kots_svrfy(&a, &pk, mu, &sig, profile).expect("svrfy");
        kots_wvrfy(&a, &pk, mu, &sig, profile).expect("wvrfy");

        assert!(kots_ivrfy(&a, &pk, b"different mu", &sig, profile).is_err());
    }

    #[test]
    fn d256_k4_kots_self_consistency() {
        run_self_consistency(&D256_K4);
    }

}
